lib/Runtime/Language/JavascriptOperators.cpp

```diff
@@ -1739,7 +1739,7 @@ namespace Js
             //         Also we might want to throw here instead of checking it again in the caller
             if (value && !requestContext->IsUndeclBlockVar(*value) && !WithScopeObject::Is(object))
             {
-                CacheOperators::CachePropertyRead(instance, object, isRoot, propertyId, false, info, requestContext);
+                CacheOperators::CachePropertyRead(isSuper ? propertyObject : instance, object, isRoot, propertyId, false, info, requestContext);
```

Comment:
Besides super property access, other cases where propertyObject != instance are following:
1) instance is proxy, propertyObject is target object
2) instance is number, propertyObject is Number.prototype
3) instance is boolean, propertyObject is Boolean.prototype

By design, 1) does not cache property access (confirmed) because underlying target object can mutate without proxy knowing it.

2) and 3) are justified to cache on their prototype (propertyObject).


---
In reply to: [130143456](https://github.com/Microsoft/ChakraCore/pull/3336#discussion_r130143456) [](ancestors = 130143456)